### DISCUSSION TITLE
Waldo: a robotic manipulator utility turret

### DISCUSSION THREAD

**Discussion by mLuby** [FakeFishGames/Barotrauma#7011] [Ideas and suggestions]
# Waldo Turret

TL;DR: add a waldo turret to expand gameplay and _dramatically_ expand sub building. Requires source code changes.

> A remote manipulator, also known as a telemanipulator or **waldo** is a device which, through electronic, hydraulic, or mechanical linkages, allows a hand-like mechanism to be controlled by a human operator. The purpose of such a device is usually to move or manipulate hazardous materials for reasons of safety. -https://en.wikipedia.org/wiki/Remote_manipulator

| Real-world | examples | of waldos: |
|--|--|--|
| ![](https://i1.wp.com/asiatimes.com/wp-content/uploads/2019/12/CGTN.png?resize=932%2C546&ssl=1) | ![](https://www.starnewsonline.com/storyimage/NC/20190717/NEWS/190718602/AR/0/AR-190718602.jpg) | ![](https://ychef.files.bbci.co.uk/976x549/p034dtrc.jpg) |

## In-Game Functionality
At its core, a waldo would:
• be a **turret installation** with the following **pins**: `power_in`, `periscope_in`, `trigger_in`.
• have a **single item slot** which can **hold _any_ item**. The item remains **visible while contained**. Ideally the turret UI shows the contained item and its condition the way it does for ammo boxes.
• **use, activate, fire, or throw\* the item** as if it's being aimed by a character with **0 skill**. So a human will always be faster, more efficient, or more accurate than a waldo; sometimes safety or automation is worth that tradeoff. \*Thrown items just activate and drop.
• **rotate only**, as current turrets do—no extending needed. The contained item should be positioned offset from the turret rotation axis so that the item clearly moves when rotated.

v2:
• Add a `set_grasp_release` pin which drops the held item if `set_grasp` receives a `0` and picks up an item under the waldo's grasper area if `set_grasp` receives a `1` and the waldo isn't already holding an item.

v3:
• Add a `set_extract_insert` pin for which a `1` extracts an item (no control over which) from the _container_ closest to the waldo's empty grasper and a `0` places the held item inside the nearest container if possible (otherwise the item remains held). Note: since many items are also containers, separate `set_grasp_release` and `set_extract_insert` pins keep it clear whether the intent is to pick up a crate or to add/remove something to/from a crate.

## Example Uses
• hull inspection and repair drone: waldo with a welder attached.
• wreck exploration drone: waldo with a plasma cutter to slice into the wreck before sending in divers.
• mining drone: two waldos: one holding a plasma cutter for mining and the other empty for picking up the dropped ore or plants and placing them into a container.

## Gameplay value
It'd also have TONS of value for sub builders, for example:
🧯 Want fire suppression? Waldo holding extinguisher.
🗳️ Want to have an auto-reloader? Waldo holding an ammo box.
☢️ Want to pretend the reactor area has lethal radiation? Waldo holding fuel rods.
🔥 Want a a flamethrower turret to burn out xenos? Waldo holding flamer. 

## Implementation
I implemented the basic version—a turret that can hold any item and use/fire it—via a few source code changes about a year ago, so I know it's possible (and pretty fun actually). But I've since lost the branch. 😅 I haven't tried picking up or dropping items; that might not be too hard if the items are lying about, but dropping them into a container might be more difficult, and picking something out of a container could be really challenging to implement—I'm just not sure.

Related ideas:
• A "payload specialist" [talent tree], for players who like to focus on remote manipulation. Perhaps that gives the waldo >0 skill, maybe it just increases rotation speed like weapons skill does for gun turrets. Mechanics could be a good job class for this.
• Maybe instead of just dropping, thrown items receive speed and direction based on a _rotating_ waldo's angular velocity.
• waldos would rarely be 100% reusable since most use/fire-type items consume a contained resource (welder fuel, oxygen tank, battery, ammunition), meaning the waldo may still need manual servicing now and then.

Thanks for reading!
